File: typescript/packages/bench/src/support/abstractReport.ts
/** Base shared by all 24 reports. */

import type { CacheableContract } from "../contracts/cacheableContract";
import type { ReportContract, ReportRow } from "../contracts/reportContract";
import type { Dataset } from "../dataset";
import { DEFAULT_CACHE_TTL_SECONDS, type HasCache } from "../concerns/hasCache";
import { AbstractComponent } from "./abstractComponent";

export abstract class AbstractReport
    extends AbstractComponent
    implements ReportContract, CacheableContract, HasCache
{
    static readonly DEFAULT_DECIMALS = 2;

    readonly decimals = AbstractReport.DEFAULT_DECIMALS;
    readonly cacheNamespace = "reports";

    /** Every subclass renders its own body; everything else is inherited. */
    abstract rows(data: Dataset): ReportRow[];

    total(data: Dataset): number {
        return this.rows(data).reduce((sum, row) => sum + row.value, 0);
    }

    isEmpty(data: Dataset): boolean {
        return this.rows(data).length === 0;
    }

    cacheKey(): string {
        return this.cacheKeyFor(this.slug);
    }

    cacheKeyFor(suffix: string): string {
        return `${this.cacheNamespace}:${suffix}`;
    }

    cacheTtl(): number {
        return DEFAULT_CACHE_TTL_SECONDS;
    }

    ttlSeconds(): number {
        return this.cacheTtl();
    }

    isCacheable(): boolean {
        return this.ttlSeconds() > 0;
    }
}

File: rust/crates/atelier-bench/src/support/abstract_report.rs
//! Base shared by all 24 reports.

use crate::concerns::has_cache::HasCache;
use crate::contracts::cacheable_contract::CacheableContract;
use crate::support::abstract_component::ComponentBase;

/// Identity plus rendering precision.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ReportBase {
    pub component: ComponentBase,
    pub decimals: u8,
}

impl ReportBase {
    pub const DEFAULT_DECIMALS: u8 = 2;

    pub const fn new(slug: &'static str, title: &'static str) -> Self {
        Self {
            component: ComponentBase::new(slug, title),
            decimals: Self::DEFAULT_DECIMALS,
        }
    }

    pub const fn slug(&self) -> &'static str {
        self.component.slug
    }

    pub const fn title(&self) -> &'static str {
        self.component.title
    }
}

impl HasCache for ReportBase {
    fn cache_namespace(&self) -> &'static str {
        "reports"
    }
}

impl CacheableContract for ReportBase {
    fn cache_key(&self) -> String {
        self.cache_key_for(self.slug())
    }
}
